src/vm/functions/iterables.rs

```diff
@@ -0,0 +1,205 @@
+    let iterable = eval(&args[1], env, context)?;
+
+    match iterable {
+        Value::List(mut list) => {
+            let mut filtered_vec = Vec::new();
+            for x in list.data.drain(..) {
+                let argument = [ SymbolicExpression::atom_value(x.clone()) ];
+                let filter_eval = apply(&function, &argument, env, context)?;
+                if let Value::Bool(include) = filter_eval {
+                    if include {
+                        filtered_vec.push(x);
+                    } // else, filter out.
+                } else {
+                    return Err(CheckErrors::TypeValueError(BoolType, filter_eval).into())
+                }
+            }
+            Value::list_with_type(filtered_vec, list.type_signature)
+        },
+        Value::Buffer(mut buff) => {
+            let mut filtered_vec = Vec::new();
+            for x in buff.data.drain(..) {
+                let v = Value::buff_from(vec![x.clone()]).unwrap();
+                let argument = [ SymbolicExpression::atom_value(v) ];
+                let filter_eval = apply(&function, &argument, env, context)?;
+                if let Value::Bool(include) = filter_eval {
+                    if include {
+                        filtered_vec.push(x);
+                    } // else, filter out.
+                } else {
+                    return Err(CheckErrors::TypeValueError(BoolType, filter_eval).into())
+                }
+            }
+            Value::buff_from(filtered_vec)
+        },
+        _ => Err(CheckErrors::ExpectedListOrBuffer(TypeSignature::type_of(&iterable)).into())
+    }
+}
+
+pub fn native_fold(args: &[SymbolicExpression], env: &mut Environment, context: &LocalContext) -> Result<Value> {
+    check_argument_count(3, args)?;
+
+    let function_name = args[0].match_atom()
+        .ok_or(CheckErrors::ExpectedName)?;
+
+    let function = lookup_function(&function_name, env)?;
+    let iterable = eval(&args[1], env, context)?;
+    let initial = eval(&args[2], env, context)?;
+
+    match iterable {
+        Value::List(mut list) => {
+            list.data.drain(..).try_fold(initial, |acc, x| {
+                let arguments = vec![
+                    SymbolicExpression::atom_value(x), 
+                    SymbolicExpression::atom_value(acc)];
+                apply(&function, &arguments, env, context)
+            })
+        },
+        Value::Buffer(mut buff) => {
+            buff.data.drain(..).try_fold(initial, |acc, x| {
+                let arguments = vec![
+                    SymbolicExpression::atom_value(Value::buff_from(vec![x]).unwrap()), 
+                    SymbolicExpression::atom_value(acc)];
+                apply(&function, &arguments, env, context)
+            })
+        },
+        _ => Err(CheckErrors::ExpectedListOrBuffer(TypeSignature::type_of(&iterable)).into())
+    }
+}
+
+pub fn native_map(args: &[SymbolicExpression], env: &mut Environment, context: &LocalContext) -> Result<Value> {
```

Comment:
It'd be good to dedupe these `map`, `filter`, `fold`, `concat`, and `assert_max_len` functions a little bit -- there's a lot of repeated code for buffers and lists, that I think can be cut down a little bit.